pkg/cvo/upgradeable.go

```diff
@@ -234,6 +247,137 @@ func (check *clusterManifestDeleteInProgressUpgradeable) Check() *configv1.Clust
 	return nil
 }
 
+func gateApplicableToCurrentVersion(gateName string, currentVersion string) (bool, error) {
+	var applicable bool
+	if ackVersion := adminAckGateRegexp.FindString(gateName); ackVersion == "" {
+		return false, fmt.Errorf("%s configmap gate name %s has invalid format; must begin with 'ack-X.Y'.",
+			internal.AdminGatesConfigMap, gateName)
+	} else {
+		parts := strings.Split(ackVersion, "-")
+		ackMinor := clusterversion.GetEffectiveMinor(parts[1])
+		cvMinor := clusterversion.GetEffectiveMinor(currentVersion)
+		if ackMinor == cvMinor {
+			applicable = true
+		}
+	}
+	return applicable, nil
+}
+
+func checkAdminGate(gateName string, gateValue string, currentVersion string,
+	ackConfigmap *corev1.ConfigMap) (string, string) {
+
+	if applies, err := gateApplicableToCurrentVersion(gateName, currentVersion); err == nil {
+		if !applies {
+			return "", ""
+		}
+	} else {
+		klog.Error(err)
+		return "AdminAckConfigMapGateNameError", err.Error()
+	}
+	if gateValue == "" {
+		message := fmt.Sprintf("%s configmap gate %s must contain a non-empty value.", internal.AdminGatesConfigMap, gateName)
+		klog.Error(message)
+		return "AdminAckConfigMapGateValueError", message
+	}
+	if val, ok := ackConfigmap.Data[gateName]; !ok || val != "true" {
+		return "AdminAckRequired", gateValue
+	}
+	return "", ""
+}
+
+type clusterAdminAcksCompletedUpgradeable struct {
+	adminGatesLister listerscorev1.ConfigMapNamespaceLister
+	adminAcksLister  listerscorev1.ConfigMapNamespaceLister
+	cvLister         configlistersv1.ClusterVersionLister
+	cvoName          string
+}
+
+func (check *clusterAdminAcksCompletedUpgradeable) Check() *configv1.ClusterOperatorStatusCondition {
+	cv, err := check.cvLister.Get(check.cvoName)
+	if meta.IsNoMatchError(err) || apierrors.IsNotFound(err) {
+		message := fmt.Sprintf("Unable to get ClusterVersion, err=%v.", err)
+		klog.Error(message)
+		return &configv1.ClusterOperatorStatusCondition{
+			Type:    configv1.ClusterStatusConditionType("UpgradeableAdminAckRequired"),
```

Comment:
We have two ~internal type constants already:

```console
$ git --no-pager grep ClusterStatusConditionType pkg/cvo/status.go 
pkg/cvo/status.go:      ClusterStatusFailing = configv1.ClusterStatusConditionType("Failing")
pkg/cvo/status.go:const ClusterVersionInvalid configv1.ClusterStatusConditionType = "Invalid"
```

Can we follow one of those approaches for `UpgradeableAdminAckRequired` too?  That way the compiler will catch us if we accidentally typo the string in of of these consumers.

Reply:
Done